metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -409,6 +410,15 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
         for (BrokerRegistrationRequestData.Feature feature : request.features()) {
             record.features().add(processRegistrationFeature(brokerId, finalizedFeatures, feature));
         }
+        if (request.features().find(KRaftVersion.FEATURE_NAME) == null) {
```

Comment:
This shouldn't be necessary. We should just leave out the `kraft.version` if it it is 0.

Reply:
> We should just leave the kraft.version out of the RPCs and records if it is 0.
Agree with this statement you made offline, I'm taking this github comment to mean let's not bother adding kraft.version in the record if it's 0, but let's keep the kraft.version check (e.g. to catch the case we have a kraft controller with finalized version 1, and a broker registration request comes in w/o kraft.version in its supported features. We can only assume the broker is on kraft.version 0 and should reject the request.)

Reply:
I see, iterating over _the controller features_ and checking brokers can support them, vs iterating over the broker features.